pycaw/magic.py

```diff
@@ -0,0 +1,563 @@
+
+        # keep reference to magic_app to check later
+        # if new session should be added to this magic_app
+        cls.magic_apps.append(magic_app)
+        log.info(f"added {magic_app} to watchlist ({len(cls.magic_apps)})")
+
+    @classmethod
+    def remove_session(cls, iid, magic_app=None):
+        """magic_root_session will get removed because it is expired"""
+        magic_root_session = cls.magic_root_sessions.pop(iid, None)
+
+        # unregister callback
+        # (magic_root_session -> MagicRootSession -> IAudioSessionEvents)
+        if magic_root_session:
+            magic_root_session.unregister_notification()
+
+        # delete iid also from magic_app or magic_session
+        # TODO: add remove handle callback?
+        cls.magic_sessions.pop(iid, None)
+        if magic_app:
+            magic_app.magic_root_sessions.pop(iid, None)
+
+    @classmethod
+    def unregister_all(cls):
+        # since cls.active can always change (multithreading)
+        # instead of a for loop i present you the while loop:
+        # while cls.active contains any items,
+        # they get popped and unregistered
+        while cls.magic_root_sessions:
+            # ________ REMOVES 1 ITEM FROM LIST ________
+            _, magic_root_session = cls.magic_root_sessions.popitem()
+
+            try:
+                # could raise an error if a session is closed
+                # between popitem() and unregister_notification()
+                magic_root_session.unregister_notification()
+                log.info(":: :: :: unregister Session")
+            except COMError as e:
+                # caused in pycaw.utils.AudioSession
+                # self._ctl.UnregisterAudioSessionNotification(self._callback)
+                # 'element not found'
+                warnings.warn(
+                    f"\nSession and Mixer closed simultaneous?\n{str(e)}")
+
+
+def for_sessions(func):
+    """Decorator for looping through sessions in MagicApp."""
+    def wrapper(self, *args):
+        if self.magic_root_sessions:
+            rv = [
+                func(self, magic_root_session, *args)
+                for magic_root_session in self.magic_root_sessions.values()
+                ]
+            # return if multiple sessions the one
+            # with mute true or the highest volume
+            rv = max(rv)
+            return rv
+        else:
+            return None
+    return wrapper
+
+
+class MagicApp():
+    """
+    When instantiated with atleast one app_exe name,
+    will be able to get/ set volume etc, also if the
+    session is created after initialize.
+    """
+    guid = pointer(GUID('{E0BD1A40-9624-44FC-A607-2ED4F00B1CC4}'))
+
+    def __init__(self,
+                 app_exe,
+                 volume_callback=None,
+                 mute_callback=None,
+                 state_callback=None):
+
+        # normalize app_exe
+        if type(app_exe) not in (tuple, list):
+            app_exe = (app_exe,)
```

Comment:
Curious, why not making it a set? It should be unique, plus the search would become ~O(1) rather than ~O(n)

Reply:
thats a great idea! 😄